py4ai/core/data/layer/v1/pandas/repository.py

```diff
@@ -0,0 +1,222 @@
+
+        """
+        return (await self.save([obj]))[0]
+
+    def __create_series_from_entity(self, entity: E) -> pd.Series:
+        serie = self.serializer.to_document(entity)
+        serie.name = self.serializer.to_document_key(self.serializer.get_key(entity))
+        return serie
+
+    async def save(self, entities: Sequence[E]) -> Sequence[E]:
+        """
+        Insert many objects of type Document/pd.DataFrame/pd.Series in a pd.DataFrame.
+
+        :param objs: List of objects to be inserted. The objects can be of the following class instances
+            :class:`cgnal.data.model.text.Document`, `pd.DataFrame` or `pd.Series`
+        :return: self i.e. an instance of ``PandasArchiver`` with updated self.data object
+
+        """
+        new = pd.DataFrame(
+            [self.__create_series_from_entity(entity) for entity in entities]
+        )
+        self.data = pd.concat(
+            [self.data.loc[list(set(self.data.index).difference(new.index))], new]
+        ).sort_index()
+        self.commit()
+        return entities
+
+    async def list(self, options: QueryOptions = QueryOptions()) -> Paged[E]:
+        return await self.retrieve_by_criteria(PandasSearchCriteria.empty(), options)
+
+    async def delete(self, key: KE) -> bool:
+        document_key = self.serializer.to_document_key(key)
+        self.data = self.data.drop(document_key)
+        self.commit()
+        return True
+
+    async def delete_by_criteria(self, criteria: SearchCriteria[Q]) -> bool:
+        rows = self.data[criteria.query(self.data)]
+        self.data = self.data.drop(rows.index)
+        self.commit()
+        return True
+
+
+class CsvRepository(PandasRepository[KE, KD, E]):
+    """Archiver based on persistent layers based on tabular files stored on csv."""
+
+    def __init__(
+        self,
+        filename: PathLike,
+        serializer: DataSerializer[KE, KD, E, pd.Series],
+        sep: str = ";",
+    ) -> None:
+        """
+        Create an in-memory archiver based on structured data stored in the filesystem as a CSV.
+
+        :param filename: str, path object or file like object. Any valid string path to a csv file.
+        :param serializer: An instance of serializer to convert between raw and domain objects.
+        :param sep: str, default ';'. Delimiter to use
+        """
+        super(CsvRepository, self).__init__(serializer)
+
+        self.filename = filename
+        self.sep = sep
+
+    def _write(self) -> None:
+        """Write object to a csv file."""
+        self.data.to_csv(self.filename, sep=self.sep)
+
+    def _read(self) -> pd.DataFrame:
+        """
+        Read csv file into a pandas DataFrame.
+
+        :return: pandas Dataframe
+        """
+        try:
+            output = pd.read_csv(self.filename, sep=self.sep, index_col=0)
+        except (FileNotFoundError, EmptyDataError):
+            output = pd.DataFrame()
+        return output
```

Comment:
Why isn't there a PicklePandasRepository too?

Reply:
We have never implemented it in other projects. We can probably create the PickleRepo later in a dedicated task